### PR TITLE
#161779789 Add User Notification Feature

### PR DESCRIPTION
**Pull request by chinemelu**
#### What does this PR do?
Adds `user email and in app notification` functionality

#### Description of Task to be completed?
* Write integration tests for `GET/api/v1/notifications/email` and `DELETE/api/v1/notifications/app`   endpoints
* Create `GET/api/v1/notifications/email` and `DELETE/api/v1/notifications/app`  endpoints
* Create `notification` migrations file and model file
* Create notification middleware and helper functions

#### How should this be manually tested?
After cloning the repo, cd into it, run `npm install` and run `npm test`
* Using Postman, test every endpoint above with this header: 
key: Content-Type value: application/json
* To test authentication, add this to the header: Get token from signup endpoint
key: 'x-access-token'  value: token
* Only authenticated users with tokens can gain access to this route

#### Any background context you want to provide?
NA

#### What are the relevant pivotal tracker stories?
[#161779789](https://www.pivotaltracker.com/story/show/161779789)

#### Screenshots (if appropriate)
NA 

#### Questions:
NA

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -29,6 +29,7 @@
     "babel-preset-env": "1.7.0",
     "babel-register": "6.26.0",
     "bcrypt": "3.0.2",
+    "bluebird": "^3.5.3",
```
Comment (Daymorelah) [inline package.json]: Lock the version of the "Bluebird" package you installed. See line 32 and line 60

File: public/index.html [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+<html>
+  <head>
+    <title>Testing socket connection</title>
+  </head>
+  <body>
+  <h1>Testing Socket Connection</h1>
+  <script src='https://cdnjs.cloudflare.com/ajax/libs/socket.io/2.2.0/socket.io.dev.js'></script>
+  <script src='./index.js'></script>
+  </body>
+</html>
```
Comment (Daymorelah) [inline public/index.html]: The title tag of this HTML document is saying "Testing socket connection". Will you be using this HTML document in production?. If we will not be using it, then I don't think this HTML document should be included in this pull request.

File: src/app.js [hunk 3]
```diff
@@ -5,13 +5,33 @@ import cors from 'cors';
...
+  socket.on('inAppNotifications', (data) => {
+    socket.emit('inAppNotifications', data);
+  });
+});
+
+const ioMiddleware = (req, res, next) => {
+  req.io = io;
+  next();
+};
+
```
Comment (Daymorelah) [inline src/app.js:33]: I think this middleware definition, from line 29, should be in the middleware folder and not in this file so that the application can be modular, easily maintainable, and easy to test.

File: src/app.js [hunk 4]
```diff
@@ -5,13 +5,33 @@ import cors from 'cors';
...
+});
+
+const io = socketIO(server);
+
+io.on('connection', (socket) => {
+  socket.on('inAppNotifications', (data) => {
+    socket.emit('inAppNotifications', data);
+  });
+});
+
```
Comment (Daymorelah) [inline src/app.js:28]: I think the setup for socket.io, see line 23, should be in its own file and imported into the app.js file so that in the nearest future when we need to add more configurations or need to implement other features on the app that needs socket.io to be re-configured, we will work in that file and not in the app.js file. This will make our code more modular.

File: src/app.js [hunk 5]
```diff
@@ -5,13 +5,33 @@ import cors from 'cors';
...
 import routes from './routes';
 
 // Create global app object
 const app = express();
 
+// let's start our server...
+const server = app.listen(process.env.PORT || 4001, () => {
+  console.log(`Server is up and running on port ${server.address().port}...!`); //eslint-disable-line
+});
+
```
Comment (Daymorelah) [inline src/app.js:20]: I feel we should start our server after we have added all necessary middlewares and routes for the app. I feel starting our server should be the last thing/logic in our app.js file. It's just a suggestion.

File: src/controller/NotificationController.js [hunk 6]
```diff
@@ -0,0 +1,116 @@
+import dotenv from 'dotenv';
+import NotificationUtil from '../utilities/Notification';
+
+dotenv.config();
+
+
+/**
+ * @description class for controlling notifications
+ */
```
Comment (Daymorelah) [inline src/controller/NotificationController.js:9]: I feel a good description will be "@description class definition for handling notifications". I think its clearer and easier to understand than the one you wrote on line 8.

File: src/controller/NotificationController.js [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+class NotificationController {
+  /**
+   *@description opts into email notification
+
```
Comment (Daymorelah) [inline src/controller/NotificationController.js:13]: I feel "subscribe for email notifications" is clearer and easier to understand. see line 12.

File: src/controller/NotificationController.js [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+      .updateNotificationStatus({ emailNotification: true }, req.user);
+    res.status(200).json({
+      success: true,
+      message: 'You have subscribed for email notifications'
+    });
+  }
+
+  /**
+   *@description opts out of email notifications
+
```
Comment (Daymorelah) [inline src/controller/NotificationController.js:39]: I feel "unsubscribe for email notifications" is better for the description. see line 38.

File: src/controller/NotificationController.js [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+  }
+
+  /**
+   *@description opts into in app notifications
+
+   * @param {Object} req - Http request
+   * @param {Object} res - Http response
+   *
+   * @returns {Object} - database response
+   */
```
Comment (Daymorelah) [inline src/controller/NotificationController.js:70]: Your code returns a response object but your documentation says database response. Ditto all the JSDoc comments for your class methods. Please document your class method well.

File: package.json [hunk 10]
```diff
@@ -56,6 +57,7 @@
     "sequelize": "4.41.2",
     "sequelize-cli": "5.3.0",
     "slug": "0.9.2",
+    "socket.io": "^2.2.0",
```
Comment (ah-med) [inline package.json]:
```suggestion
    "socket.io": "2.2.0",
```

**Review by Daymorelah — CHANGES_REQUESTED** [review]

File: src/controller/NotificationController.js [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+ */
+class NotificationController {
+  /**
+   *@description opts into email notification
+
+   * @param {Object} req - Http request
+   * @param {Object} res - Http response
+   *
+   * @returns {boolean} - true or false
+   */
```
Comment (Daymorelah) [inline src/controller/NotificationController.js:18]: The method "optIntoEmailNotifications" returns an HTTP response object but your JSDoc comment says returning a boolean (see line 17). Please update your JSDoc to have the correct return type.

File: src/controller/NotificationController.js [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+  static async optIntoEmailNotifications(req, res) {
+    const isUserEmailNotificationOn = await NotificationUtil
+      .isUserEmailNotificationOn(req.user);
+
+    if (isUserEmailNotificationOn) {
+      return res.status(400).json({
+        success: false,
+        message: 'You are already subscribed for email notifications'
+      });
```
Comment (Daymorelah) [inline src/controller/NotificationController.js:27]: I think "You have already subscribed for email notifications" is the right sentence you are trying to send to the user.

File: src/controller/NotificationController.js [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+  }
+
+  /**
+   *@description opts out of email notifications
+
+   * @param {Object} req - Http request
+   * @param {Object} res - Http response
+   *
+   * @returns {Object} - Database response
+   */
```
Comment (Daymorelah) [inline src/controller/NotificationController.js:44]: The method "optOutOfEmailNotifications", (line 54), returns an HTTP response object but your JSDoc comment says returning a database response (see line 17). Please update your JSDoc to have the correct return type.

File: src/controller/articlesController.js [hunk 14]
```diff
@@ -216,11 +240,43 @@ class ArticleController {
...
+        return res.status(404).json({ message: 'User does not exist' });
+      }
+
+      const isValidId = await uuidV4Validator(articleId);
+
+      if (!isValidId) {
+        return res.status(400).json({
+          success: false,
+          message: 'Invalid Id'
+        });
```
Comment (Daymorelah) [inline src/controller/articlesController.js]: The message field of the HTTP response object on line 258 does not contain a self-explanatory message. "Invalid article ID" would have been more descriptive.

File: src/controller/articlesController.js [hunk 15]
```diff
@@ -216,11 +240,43 @@ class ArticleController {
...
+      }
+
+      if (bookmark) {
+        return res.status(409).json({
+          message: 'Bookmark already exists'
+        });
+      }
+
+      const name = req.body.name || article.dataValues.title;
+
```
Comment (Daymorelah) [inline src/controller/articlesController.js:251]: This controller "bookmarkArticle" does a bunch of things in its definition like checking if there is a valid user in the token(lines 247 & 249), checking if an article ID is valid(line 253 & 255), checking if an article and bookmark exist(line 262 - 276). All the logic related to validating these things should be in a separate file. Your current implementation of "bookmarkArticle" does not follow the Single Responsibility Principle (SRP). This will make the controller difficult to maintain. The "bookmarkArticle" should only bookmark an article and nothing more.

File: src/controller/userController.js [hunk 16]
```diff
@@ -147,15 +154,78 @@ class UserController {
...
+        });
+      }
+
+      const { username, email } = doesUserInParamsExist;
+      const details = {
+        email,
+        subject: 'Author\'s Haven - Email notification',
+        emailBody: `<p>You are now being followed by ${username}</p>`
+      };
+
```
Comment (Daymorelah) [inline src/controller/userController.js:164]: This controller "followUser" does not follow the Single Responsibility Principle (SRP). The controller contains lines of code that does other functions other than the logic needed to follow a user. The code for validating a user ID, checking if a user has already followed another user, e.t.c should be in a separate method or file and not in the "followUser" controller.

File: src/controller/userController.js [hunk 17]
```diff
@@ -147,15 +154,78 @@ class UserController {
...
+      }
+
+      const isValidId = await uuidV4Validator(userId);
+
+      if (!isValidId) {
+        return res.status(400).json({
+          success: false,
+          message: 'Invalid Id'
+        });
+      }
```
Comment (Daymorelah) [inline src/controller/userController.js]: Please make your message more descriptive (see line 172). "Invalid userID" is more descriptive.

File: src/controller/userController.js [hunk 18]
```diff
@@ -176,24 +246,50 @@ class UserController {
...
+          message: 'You have unfollowed this user'
+        });
+      } else {
+        res.status(400).json({
+          success: false,
+          message: 'You do not follow this user'
+        });
+      }
+    } catch (error) {
+      throw error;
```
Comment (Daymorelah) [inline src/controller/userController.js]: Why do you prefer to throw this error, (line 292), and not handle it by sending a response to the user when an error is caught in the catch block?.

File: src/controller/userController.js [hunk 19]
```diff
@@ -176,24 +246,50 @@ class UserController {
...
     }
```
Comment (Daymorelah) [inline src/controller/userController.js:211]: Please use the feedback given on the "followUser" controller to make this controller follow the Single Responsibility Principle. Put all unrelated code in a separate method or file.

File: src/middleware/Authorization.js [hunk 20]
```diff
@@ -79,20 +79,17 @@ class Authorization {
...
-      } catch (error) {
-        return res.status(401).send({
-          success: false,
+        res.status(401).send({
+          code: 401,
           message: 'Authentication failed',
         });
+      } catch (error) {
+        throw error;
       }
```
Comment (Daymorelah) [inline src/middleware/Authorization.js:94]: Why did you decide to throw the error instead of handling it by sending a response to the user?

File: src/middleware/Authorization.js [hunk 21]
```diff
@@ -79,20 +79,17 @@ class Authorization {
...
-          });
+          return next();
         }
-      } catch (error) {
-        return res.status(401).send({
-          success: false,
+        res.status(401).send({
+          code: 401,
           message: 'Authentication failed',
         });
```
Comment (Daymorelah) [inline src/middleware/Authorization.js]: This method "checkToken" has unnecessary code changes. You should have consulted with the author of the code before modifying the code.

File: src/middleware/Authorization.js [hunk 22]
```diff
@@ -79,20 +79,17 @@ class Authorization {
...
         if (tokenVerified.success) {
           req.decoded = tokenVerified;
-          next();
-        } else {
-          return res.status(401).send({
-            success: false,
-            message: 'Authentication failed',
-          });
+          return next();
         }
```
Comment (Daymorelah) [inline src/middleware/Authorization.js:91]: Express automatically calls return when you call the next method. The return keyword is unnecessary

File: src/middleware/Authorization.js [hunk 23]
```diff
@@ -120,5 +117,28 @@ class Authorization {
       message: 'Invalid Id'
     });
```
Comment (Daymorelah) [inline src/middleware/Authorization.js:119]: Please make your response message more descriptive. What ID is being invalid?. Is it a userID, articleID, bookmarkID, e.t.c.

**Review by Daymorelah — CHANGES_REQUESTED** [review]

File: src/utilities/Notification.js [hunk 24]
```diff
@@ -0,0 +1,144 @@
...
+    try {
+      if (user.inAppNotification) {
+        return Notification.create({
+          userId: user.id,
+          text
+        });
+      }
+    } catch (error) {
+      throw error;
+    }
```
Comment (Daymorelah) [inline src/utilities/Notification.js:35]: The throw statement throws a user-defined exception. check this link on when and how to use the "throw" keyword. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/throw. Try to handle the error being caught in the catch block properly. I noticed you were throwing errors in all your implementation. Try to make appropriate adjustments.

**Review by Daymorelah — CHANGES_REQUESTED** [review]

File: src/utilities/Notification.js [hunk 25]
```diff
@@ -0,0 +1,144 @@
...
+   * @param {Array} userArray - The array of users
+   * @param {string} text - The text of the notification
+   *
+   * @returns {object} sends database query result
+   */
+  static async setMultipleAppNotifications(userArray, text) {
+    try {
+      return models.sequelize
+        .transaction(t => Promise.map(userArray, (user) => {
+          if (user.inAppNotification) {
```
Comment (Daymorelah) [inline src/utilities/Notification.js:51]: The letter 't' on line 46 is a poor choice for a varaible name. Please use a more descriptive variable name.

File: src/utilities/checkArticleExists.js [hunk 26]
```diff
@@ -0,0 +1,34 @@
...
+        articleId,
+        {
+          include:
+          {
+            model: Users,
+            attributes: ['email']
+          }
+        }
+      );
+    return article;
```
Comment (Daymorelah) [inline src/utilities/checkArticleExists.js:37]: I think there is an agreed format for a response i.e. using "success: false/true" and an optional message. I think returning the exact response from the database is not good enough. You should either return a boolean or a more descriptive response.

File: src/utilities/checkBookmarkExists.js [hunk 27]
```diff
@@ -0,0 +1,33 @@
...
+  try {
+    const bookmark = await Bookmark.findOne({
+      where: {
+        [Op.and]: [{ isActive: true },
+          { userId },
+          { articleId }
+        ]
+      }
+    });
+    return bookmark;
```
Comment (Daymorelah) [inline src/utilities/checkBookmarkExists.js]: Ditto!!!

File: src/utilities/findDatabaseField.js [hunk 28]
```diff
@@ -2,55 +2,61 @@ import models from '../models';
...
    */
-  static async UserInToken(req, res, next) {
-    const userId = req.decoded.id;
-    const user = await findUser(userId);
-    if (user) {
-      return next();
+  static async findUser(id) {
+    try {
+      const queryUserResult = await Users.findByPk(id);
+      return queryUserResult;
```
Comment (Daymorelah) [inline src/utilities/findDatabaseField.js:21]: Ditto!!!

File: src/utilities/followers.js [hunk 29]
```diff
@@ -63,51 +63,22 @@ class Followers {
...
    *
    * @param {object} req Http request
    * @param {object} res Http response
    * @param {function} next callback
    *
    * @returns  {JSON} Returns a JSON object
    */
-  static async checkForExistingFollowing(req, res, next) {
+  static async checkForSelfFollow(req, res, next) {
     const userId = await req.params.id;
```
Comment (Daymorelah) [inline src/utilities/followers.js]: This should be in the middleware folder not in the utility folder. Please use the folder structure the team agreed on.

File: src/utilities/followers.js [hunk 30]
```diff
@@ -122,39 +93,42 @@ class Followers {
    *
    * @returns  {JSON} Returns a JSON object
    */
-  static async checkForSelfFollow(req, res, next) {
+  static async checkForSelfUnfollow(req, res, next) {
     const userId = await req.params.id;
```
Comment (Daymorelah) [inline src/utilities/followers.js]: Ditto!!. This method "checkForSelfUnfollow" should be in the middleware folder and not in the utility folder.

File: src/utilities/isValidId.js [hunk 31]
```diff
@@ -0,0 +1,20 @@
...
+   * @returns {Boolean} Returns an object
+   */
+const uuidV4Validator = async (id) => {
+  const uuidV4Regex = new RegExp(['^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-',
+    '[89AB][0-9A-F]{3}-[0-9A-F]{12}$'].join(''), 'i');
+  const result = await uuidV4Regex.test(id);
+  if (result) {
+    return result;
+  }
+};
```
Comment (Daymorelah) [inline src/utilities/isValidId.js:24]: I think Sequelize had a UUIDV4 validator. You should check it out.

**Review by Daymorelah — DISMISSED** [review]

File: src/controller/articlesController.js (hunk 14 above)
Comment (chinemelu) [reply src/controller/articlesController.js]: I feel it's unnecessary to have a descriptive message in this instance as the user hitting the endpoint knows the id that's invalid - It's always the id in the params. Using a descriptive message in this instance would lead to complications.

File: src/middleware/Authorization.js (hunk 21 above)
Comment (chinemelu) [reply src/middleware/Authorization.js]: It wasn't unnecessary. I did consult the author before modifying the code. The `code 401` bit was included in an integration test so without that, the test fails.

File: src/middleware/Authorization.js (hunk 22 above)
Comment (chinemelu) [reply src/middleware/Authorization.js:91]: This is unfortunately not the case always. In some cases, the `return` statement is unnecessary, however, when there's a response message after that block, it becomes extremely necessary unless the middleware doesn't function as it should.

File: src/utilities/Notification.js (hunk 25 above)
Comment (chinemelu) [reply src/utilities/Notification.js:51]: The letter `t` is the default letter used in the sequelize documentation.

File: src/utilities/checkArticleExists.js (hunk 26 above)
Comment (chinemelu) [reply src/utilities/checkArticleExists.js:37]: I return the article because it's used in a different file. The function is not intended to send a response to the user. This way, anytime the function is invoked, it either returns an article (if it exists) or nothing (if the article doesn't exist). If I want to retrieve anything from the existing article, I also can.  Also, saying something is `not good enough` is not ASK feedback, for your information :( That's my feedback on your feedback.

File: src/utilities/followers.js (hunk 29 above)
Comment (chinemelu) [reply src/utilities/followers.js]: This is well noted and has been changed.

File: src/utilities/isValidId.js (hunk 31 above)
Comment (chinemelu) [reply src/utilities/isValidId.js:24]: There are instances where the sequelize validator cannot be used. This method is for those instances.

File: src/utilities/checkBookmarkExists.js (hunk 27 above)
Comment (chinemelu) [reply src/utilities/checkBookmarkExists.js]: Same as my response above.

File: src/utilities/findDatabaseField.js (hunk 28 above)
Comment (chinemelu) [reply src/utilities/findDatabaseField.js:21]: Same as my response above.

File: src/utilities/followers.js (hunk 30 above)
Comment (chinemelu) [reply src/utilities/followers.js]: Noted. I have changed it accordingly.

File: src/middleware/Authorization.js (hunk 23 above)
Comment (chinemelu) [reply src/middleware/Authorization.js:119]: I feel it's not necessary as the user knows that the invalid Id is the one on the params.

File: src/controller/NotificationController.js (hunk 8 above)
Comment (chinemelu) [reply src/controller/NotificationController.js:39]: I thought about this name, too, but it's too long.

File: src/controller/NotificationController.js (hunk 7 above)
Comment (chinemelu) [reply src/controller/NotificationController.js:13]: I thought about this name, too, but it's too long.

File: src/controller/NotificationController.js (hunk 11 above)
Comment (chinemelu) [reply src/controller/NotificationController.js]: Noted. It has been changed.

File: src/controller/NotificationController.js (hunk 12 above)
Comment (chinemelu) [reply src/controller/NotificationController.js:27]: Noted. I have changed it accordingly.

File: src/controller/NotificationController.js (hunk 13 above)
Comment (chinemelu) [reply src/controller/NotificationController.js:44]: Noted. i have changed it accordingly.

File: src/controller/NotificationController.js (hunk 11 above)
Comment (chinemelu) [reply src/controller/NotificationController.js:18]: Noted. I have changed it accordingly.

File: src/controller/NotificationController.js (hunk 6 above)
Comment (chinemelu) [reply src/controller/NotificationController.js:9]: Noted. I have changed it accordingly.

File: src/controller/NotificationController.js (hunk 9 above)
Comment (chinemelu) [reply src/controller/NotificationController.js:70]: Noted. I have changed it.

File: src/app.js (hunk 3 above)
Comment (chinemelu) [reply src/app.js:33]: Using the middleware index file will give the middleware access to all the routes and it means I don't have to include them in specific routes. It's the way Express works.

File: src/app.js (hunk 4 above)
Comment (chinemelu) [reply src/app.js:28]: You are probably right but I will leave it as it is for now.  It's definitely something to consider in any subsequent refactoring.

File: src/app.js (hunk 5 above)
Comment (chinemelu) [reply src/app.js:20]: The server variable was needed in downstream operations and the application works as it should. I will do some more consultation but I would leave it as it is for now. It's something to consider in the future.

File: src/middleware/Authorization.js (hunk 20 above)
Comment (chinemelu) [reply src/middleware/Authorization.js:94]: Noted with thanks. I have changed it accordingly.

File: src/utilities/Notification.js (hunk 24 above)
Comment (chinemelu) [reply src/utilities/Notification.js:35]: This is well noted. I have changed it accordingly.

File: src/controller/articlesController.js (hunk 15 above)
Comment (chinemelu) [reply src/controller/articlesController.js:251]: This is well noted. I thought about this as well. Thanks for bringing it to my notice. It's fixed now

File: src/controller/userController.js (hunk 19 above)
Comment (chinemelu) [reply src/controller/userController.js:211]: Thanks for the feedback. I have fixed this.

File: test/unitTest/notifications.test.js [hunk 32]
```diff
@@ -0,0 +1,152 @@
...
+  const notificationText = 'You are now following this user'
+  describe('Notification email utilities', () => {
+    let myEmailSpy;
+    beforeEach(() => {
+      myEmailSpy = sinon.spy(sendEmail, 'emailSender')
+    })
+  
+    afterEach(() => {
+      sendEmail.emailSender.restore();
+    })
```
Comment (ah-med) [inline test/unitTest/notifications.test.js:101]: I got the error below when I ran the test locally
screen shot 2019-01-06 at 11 54 02 pm
I suggest you ensure `done` is called on both hooks since they are both calling an async function `sendEmail`.
Kindly visit this [link](https://mochajs.org/#asynchronous-hooks) for more information on asynchronous hooks.

**Review by ah-med — DISMISSED** [review]

File: test/unitTest/notifications.test.js (hunk 32 above)
Comment (chinemelu) [reply test/unitTest/notifications.test.js:101]: Noted. This is annoying because the tests work locally sometimes and then don't work locally, but they always pass on Travis. I have made some changes and I think it works now. Please confirm it does.

**Review by kingisaac95 — APPROVED** [review]
